kernel/timeout.c

```diff
@@ -195,6 +158,42 @@ void z_set_timeout_expiry(s32_t ticks, bool idle)
 	}
 }
 
+void z_clock_announce(s32_t ticks)
+{
+#ifdef CONFIG_TIMESLICING
+	z_time_slice(ticks);
+#endif
+
+	k_spinlock_key_t key = k_spin_lock(&timeout_lock);
+
+	announce_remaining = ticks;
+
+	while (first() != NULL && first()->dticks <= announce_remaining) {
```

Comment:
This still allows for immediate firing of timeouts scheduled by the `t->fn(t)`, breaking monotonicity of the timer.
(please consider case when `announce_remaining` is larger than scheduled time).

Reply:
For like the ninth time, this is a feature, not a bug.  It's not the source of the problem you're seeing, I promise.

I opened #12247 to try to explain this yet again.  Please take the discussion about API changes there.

(Also: there's no violation of monotonicity -- timers expire in the order expected always.  DELTA times between timeouts will and MUST be skewed in the face of lost interrupts, the argument is which delta gets skewed.)